### PR TITLE
Add --include-non-variant-sites to GenotypeGVCFs.

### PR DESCRIPTION
**Pull request by cmnbroad**
Fixes https://github.com/broadinstitute/gatk/issues/2865.

To match GATK without sacrificing performance for existing GenotypeGVCF modes, this adds a new traversal type `VariantWalkerGroupedByLocus` which can process variants one-at-a-time, or optionally grouped by Locus. GenotypeGVCFs itself then has to further filter the variants to emulate the "prioritize start location" behavior in GATK3. If we keep this traversal, I'll add some unit tests, and we should find a better name.

In order to match GATK3 output for sites where the only alt allele is a spanning deletion, I had to use `OutputMode.EMIT_ALL_SITES` when emitting nonvariant sites only (GATK3 uses `OutputMode.EMIT_ALL_CONFIDENT_SITES` in that case). Specifically, its necessary to to prevent GenotypingEngine from entering [this](https://github.com/broadinstitute/gatk/blob/33af6de5fb990f4642245b7a80b76d251371ef09/src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java#L285) recently added code block, and short circuiting out. @ldgauthier Any opinion on whether this is too permissive ? Another option would be to special-case this, perhaps by adding another OutputMode state.

### CODE REVIEW

**Comment by ldgauthier** [comment]
@cmnbroad I haven't actually looked at the code yet, but we don't necessarily have to match GATK3 output.  In the expected file at 20:10004769 where there's a deletion and then each position within the deletion reported out again with exactly the same attributes I'm not married to that representation.  Furthermore, I don't know what SelectVariants will do if you ask for e.g. 20: 10004772 -- should you get the deletion?  The deletion and the *-only context at the requested position _without_ the previous positions in the deletion?  Does it make things any simpler if you don't have to output *-only sites?

**Comment by cmnbroad** [comment]
@ldgauthier Emitting the spanning-deletion-only sites completely would definitely make things simpler, since I could then use EMIT_ALL_CONFIDENT_SITES and GenotypingEngine would just naturally do the right thing. Also, the result would comport with my own naive expectations. Using the current PR, outputs could contain LowQual sites that GATK3 wouldn't have included. So if you're good with that, I'll update the PR and tests to reflect that before any more reviewing.

**Comment by cmnbroad** [comment]
Ok, based on internal discussions, I updated this to not emit the spanning-del only sites, but to include the LowQual sites.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                                new ReadsContext(reads, variantInterval, readFilter),
+                                new ReferenceContext(reference, variantInterval),
+                                new FeatureContext(features, variantInterval));
+
+                        progressMeter.update(variantInterval);
+                    });
+        }
+    }
+
+    protected Stream<VariantContext> getTransformedVariantStream(final Spliterator<VariantContext> source, final VariantFilter filter) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: This seems like it could maybe be pulled up to the base class and avoid duplicating it here and in multi and single variant walkers.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+     * @param referenceContext Reference bases spanning the current variant. Will be an empty, but non-null, context object
+     *                         if there is no backing source of reference data (in which case all queries on it will return
+     *                         an empty array/iterator). Can request extra bases of context around the current variant's interval
+     *                         by invoking {@link ReferenceContext#setWindow}
+     *                         on this object before calling {@link ReferenceContext#getBases}
+     * @param featureContext Features spanning the current variant. Will be an empty, but non-null, context object
+     *                       if there is no backing source of Feature data (in which case all queries on it will return an
+     *                       empty List).
+     */
+    public abstract void apply(Locatable loc, List<VariantContext> variants, ReadsContext readsContext, ReferenceContext referenceContext, FeatureContext featureContext );
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: This doc needs a big callout about the mode switching and what happens in each mode.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 3]
```diff
@@ -124,6 +125,12 @@
 
     private ReferenceConfidenceVariantContextMerger merger;
 
+    // the INFO field annotation key names to remove
+    private final List<String> infoFieldAnnotationKeyNamesToRemove = new ArrayList<>();
+
+    // INFO Header names that require alt alleles
+    final Set<String> infoHeaderAltAllelesLineNames = new LinkedHashSet<>();
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: I would type this as a LinkedHashSet if the order is important.  Otherwise someone will break it in a confusing way later.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 4]
```diff
@@ -197,32 +231,49 @@ private void setupVCFWriter(VCFHeader inputVCFHeader, SampleList samples) {
...
         }
     }
 
-
+    // If includeNonVariants is set, we're using group-by-locus traversal. To match GATK3 GenotypeGVCFs,
+    // see if there is a variant in the overlapping group that starts exactly at the locus start position, and if so
+    // prioritize and process only that variant. Otherwise process all of the overlapping variants.
+    private List<VariantContext> getVariantSubsetToProcess(final Locatable loc, List<VariantContext> preProcessedVariants) {
+        return includeNonVariants ?
+                preProcessedVariants.stream().anyMatch(vc -> vc.getStart() == loc.getStart()) ?
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Multiple nested ternary expressions with multiple stream expressions are hard to parse, and it's a bit wasteful to run the same stream twice in a row to check if something exists and then take the first one.  Could you rewrite this so it's a little clearer and can reuse the value if it finds it.  I might also add a check to be sure that there is exactly 1, because we probably have a problem if there are more than 1.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 5]
```diff
@@ -268,6 +320,56 @@ private VariantContext calculateGenotypes(VariantContext vc){
         return genotypingEngine.calculateGenotypes(vc, model, null);
     }
 
+    /**
+     * Remove INFO field annotations if no alternate alleles
+     *
+     * @param vc    the variant context
+     * @return variant context with the INFO field annotations removed if no alternate alleles
+     */
+    private VariantContext removeInfoAnnotationsIfNoAltAllele(final VariantContext vc)  {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: this is unused

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 6]
```diff
@@ -268,6 +320,56 @@ private VariantContext calculateGenotypes(VariantContext vc){
...
+
+        // If no alt allele, then remove INFO fields that require alt alleles
+        if ( newAlleles.size() == 1 ) {
+            final VariantContextBuilder builder = new VariantContextBuilder(vc).alleles(newAlleles);
+            for ( final String name : infoHeaderAltAllelesLineNames ) {
+                builder.rmAttributes(Arrays.asList(name));
+            }
+            return builder.make();
+        } else {
+            return vc;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Is it weird that this line has no test coverage?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java [hunk 7]
```diff
@@ -601,10 +603,6 @@ public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext
         }
     }
 
-    private boolean checkOnlySpanDel(VariantContext vc){
-        return vc.getAlternateAlleles().size() == 1 && vc.getAlternateAllele(0).basesMatch(Allele.SPAN_DEL);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: good find

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java [hunk 8]
```diff
@@ -171,6 +171,7 @@ their names (or descriptions) depend on some threshold.  Those filters are not i
     // Symbolic alleles
     public final static String SYMBOLIC_ALLELE_DEFINITION_HEADER_TAG = "ALT";
     public final static String NON_REF_SYMBOLIC_ALLELE_NAME = "NON_REF";
+    public final static Allele NON_REF_SYMBOLIC_ALLELE = Allele.create("<"+NON_REF_SYMBOLIC_ALLELE_NAME+">", false); // represents any possible non-ref allele at this site
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java]: This exists already `Allele.NON_REF_ALLELE`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java [hunk 9]
```diff
@@ -171,6 +171,7 @@ their names (or descriptions) depend on some threshold.  Those filters are not i
     // Symbolic alleles
     public final static String SYMBOLIC_ALLELE_DEFINITION_HEADER_TAG = "ALT";
     public final static String NON_REF_SYMBOLIC_ALLELE_NAME = "NON_REF";
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java]: Could we replace uses of this with uses of `Allele.NON_REF_STRING`.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 10]
```diff
@@ -1750,6 +1750,14 @@ public static boolean isXnp(final String reference, final String alternate) {
         return ((refComparable.length() == altComperable.length()) && (!refComparable.equals(altComperable)));
     }
 
+    /**
+     * @param vc {@link VariantContext to test}
+     * @return true if the only alternate allele for this VariantContext is a spanning deletion, otherwise false.
+     */
+    public static boolean isSpanningDeletionOnly(final VariantContext vc){
+        return vc.getAlternateAlleles().size() == 1 && vc.getAlternateAllele(0).basesMatch(Allele.SPAN_DEL);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: This should probably call `GATKVCFConstants.isSpanningDeletion`.  It's not done uniformly but it should probably be made to be uniform.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java [hunk 11]
```diff
@@ -91,8 +93,39 @@
...
-                //{getTestFile(basePairGVCF), getTestFile( "gvcf.basepairResolution.includeNonVariantSites.gatk3.7_30_ga4f720357.expected.vcf"), Collections.singletonList("--"+GenotypeGVCFs.ALL_SITES_LONG_NAME) //allsites not supported yet
+
+                // all sites/--include-non-variant-sites tests
+
+                // The results from this test differ slightly from gatk3 since sites where the only alternate allele is a spanning
+                // deletion are not emitted by gatk4.
+                {getTestFile(BASE_PAIR_GVCF), getTestFile( "expected/gvcf.basepairResolution.includeNonVariantSites.vcf"), Collections.singletonList("--" + GenotypeGVCFs.ALL_SITES_LONG_NAME), b37_reference_20_21 },
+                {getTestFile( "combine.single.sample.pipeline.1.vcf"),
+                        getTestFile( "expected/combine.single.sample.pipeline.1.include_nonvariant.vcf"),
+                        Arrays.asList( "  --include-non-variant-sites -L 20:10,030,000-10,033,000 -L 20:10,386,000-10,386,500 "),
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: I would use the constant here instead of hardcoding --include-non-variant-sites.  We're not consistent about it at all, but better to use the constants when adding new ones.  I think -L is so embedded in gatk that I wouldn't bother constantizing it...

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 12]
```diff
@@ -268,6 +320,56 @@ private VariantContext calculateGenotypes(VariantContext vc){
...
+     * @param vc   the variant context
+     * @return variant context with the NON-REF alleles removed if multiallelic or replaced with NO-CALL alleles if biallelic
+     */
+    private VariantContext removeNonRefAlleles(final VariantContext vc) {
+
+        // If NON_REF is the only alt allele, ignore this site
+        final List<Allele> newAlleles = new ArrayList<>();
+        // Only keep alleles that are not NON-REF
+        for ( final Allele allele : vc.getAlleles() ) {
+            if ( !allele.equals(GATKVCFConstants.NON_REF_SYMBOLIC_ALLELE) ) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: use the existing constant in Allele

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java [hunk 13]
```diff
@@ -47,7 +47,9 @@
     private static final File CEUTRIO_20_21_GATK3_4_G_VCF = new File(largeFileTestDir, "gvcfs/CEUTrio.20.21.gatk3.4.g.vcf");
     private static final String CEUTRIO_20_21_EXPECTED_VCF = "CEUTrio.20.21.gatk3.7_30_ga4f720357.expected.vcf";
     private static final List<String> ATTRIBUTES_TO_IGNORE = Arrays.asList(
+            "AN",
             "AS_QD",
+            "RGQ",
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: Why do we need these new exclusions?

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+/**
+ * VariantWalkerGroupedByLocus processes variants from a single source, either one at a time in order, or optionally
+ * grouped by locus overlap, with optional contextual information from a reference, sets of reads, and/or supplementary sources
+ * of Features. By-locus traversal is opt in, via {@link #setGroupByLocus()}, and only loci with overlapping
+ * variants are traversed.
+ *
+ * VariantWalkerGroupedByLocus authors must implement the {@link #apply} method to process each variant, and may optionally implement
+ * {@link #onTraversalStart}, {@link #onTraversalSuccess} and/or {@link #closeTool}.
+ */
+public abstract class VariantWalkerGroupedByLocus extends VariantWalkerBase {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Maybe `VariantLocusWalker` would be more accurate?

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+    // NOTE: using File rather than FeatureInput<VariantContext> here so that we can keep this driving source
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    //       of variants separate from any other potential sources of Features
+    @Argument(fullName = StandardArgumentDefinitions.VARIANT_LONG_NAME, shortName = StandardArgumentDefinitions.VARIANT_SHORT_NAME, doc = "A VCF file containing variants", common = false, optional = false)
+    public String drivingVariantFile;
+
+    // NOTE: keeping the driving source of variants separate from other, supplementary FeatureInputs in our FeatureManager in GATKTool
+    //we do add the driving source to the Feature manager but we do need to treat it differently and thus this field.
+    private FeatureDataSource<VariantContext> drivingVariants;
+    private FeatureInput<VariantContext> drivingVariantsFeatureInput;
+
+    private boolean traverseByLocus = false;
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+    /**
+     * Enable grouped of variants by locus. When true, the {@link #apply} method will be called with all overlapping
+     * variants at each each locus that has overlapping variants.
+     */
+    protected void setGroupByLocus() { traverseByLocus = true; }
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Inverted, and named `changeTraversalModeToByVariant`.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+     * Subclasses can override to provide their own behavior but default implementation should be suitable for most uses.
+     */
+    @Override
+    public void traverse() {
+        final CountingReadFilter readFilter = makeReadFilter();
+        final VariantFilter variantFilter = makeVariantFilter();
+
+        if (traverseByLocus) {
+            getLocusStream(getTraversalIntervals())
+                .forEach(locus -> {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Yeah, done.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+        final Iterable<SimpleInterval> iterable = () -> new IntervalLocusIterator(requestedIntervals.iterator());
+        return StreamSupport.stream(iterable.spliterator(), false);
+    }
+
+    /**
+     * Process an individual variant. Must be implemented by tool authors.
+     * In general, tool authors should simply stream their output from apply(), and maintain as little internal state
+     * as possible.
+     *
+     * @param loc
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Updated doc to reflect what loc represents.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    // Return a Stream of SimpleInterval covering the entire territory sketched out by requestedIntervals
+    private Stream<SimpleInterval> getLocusStream(final List<SimpleInterval> requestedIntervals) {
+        final Iterable<SimpleInterval> iterable = () -> new IntervalLocusIterator(requestedIntervals.iterator());
+        return StreamSupport.stream(iterable.spliterator(), false);
+    }
+
+    /**
+     * Process an individual variant. Must be implemented by tool authors.
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Fixed.

File: src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java [hunk 21]
```diff
@@ -95,6 +97,46 @@ public final VCFHeader getHeaderForVariants() {
         return drivingVariants.getHeader();
     }
 
+    /**
+     * Implementation of variant-based traversal.
+     * Subclasses can override to provide their own behavior but default implementation should be suitable for most uses.
+     */
+    @Override
+    public void traverse() {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java]: I'm going to make a separate ticket to see if we can do do mething better.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+     * Implementation of variant-based traversal.
+     * Subclasses can override to provide their own behavior but default implementation should be suitable for most uses.
+     */
+    @Override
+    public void traverse() {
+        final CountingReadFilter readFilter = makeReadFilter();
+        final VariantFilter variantFilter = makeVariantFilter();
+
+        if (traverseByLocus) {
+            getLocusStream(getTraversalIntervals())
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: I did some perf comparisons with gatk3, and this implementation was slower in the case where there are large swaths of loci with no overlapping variants. I addressed this by doing something similar to what GATK3 does, which is basically to start with locus shards (of size `FEATURE_CACHE_LOOKAHEAD`), and only drop down to per-locus for shards that have variants. Its about the same as GAtK3 now.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+                            progressMeter.update(locus);
+                        }
+                    }
+            );
+        } else {
+            // Process each variant in the input stream.
+            getTransformedVariantStream( getSpliteratorForDrivingVariants(), variantFilter )
+                    .forEach(variant -> {
+                        final SimpleInterval variantInterval = new SimpleInterval(variant);
+                        apply(variant,
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Agreed.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java (hunk 1 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Yes, done. It still takes a spliterator, though..

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java (hunk 2 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java (hunk 3 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java (hunk 4 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java (hunk 5 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: hm. @ldgauthier I think it was a mistake on my part that I included this method in this PR when I ported these changes from GATK3, since this code doesn't appear to be explicitly dependent on `includeNonVariantSites`. However, is it correct that GATK doesn't calls this ?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java (hunk 6 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: I'm using the same tests GATK3 uses, but it is a bit surprising.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java (hunk 8 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java (hunk 9 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 10 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java (hunk 11 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: Good point - done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java (hunk 12 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java (hunk 14 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: Yeah, though `VariantWalkerGroupedByLocus` has symmetry with `MultiVariantWalkerGroupedOnStart`, for what thats worth. Happy to change it if you think `VariantLocusWalker` is better .

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java (hunk 13 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: Good question. GATK4 doesn't appear to use `AN`, nor have code that [moves GQ to RGQ](https://github.com/broadinstitute/gsa-unstable/blob/c707a5d4d55da9add817793b98b4d1e089151414/protected/gatk-tools-protected/src/main/java/org/broadinstitute/gatk/tools/walkers/variantutils/GenotypeGVCFs.java#L483) like GATK3 did. So these are removed from the comparison with GATK3 output. @ldguathier - is is correct that GATK4 doesn't do this ?

Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: I'm not sure anyone actually used RGQ.  We probably didn't port that feature.

I don't know why we're ignoring AN though.  That should be part of the ChromosomeCounts annotation (also AC and AF) that's part of the standard group.  What happens if you don't ignore AN?

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+##contig=<ID=20,length=63025520,assembly=b37>
+##contig=<ID=21,length=48129895,assembly=b37>
+##reference=file:///Users/cnorman/projects/gatk/src/test/resources/large/human_g1k_v37.20.21.fasta
+##source=GenotypeGVCFs
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	GTEX-OXRP-0003	GTEX-QXCU-0004	GTEX-RVPV-0003
+20	10012730	.	G	.	27.02	.	DP=4	GT:AD:DP:RGQ	0/0:2,0:2:6	./.:0,0:0	0/0:2,0:2:6
+20	10012731	.	C	.	27.02	.	DP=4	GT:AD:DP:RGQ	0/0:2,0:2:6	./.:0,0:0	0/0:2,0:2:6
+20	10012732	.	A	.	27.02	.	DP=4	GT:AD:DP:RGQ	0/0:2,0:2:6	./.:0,0:0	0/0:2,0:2:6
+20	10012733	.	G	.	27.02	.	DP=4	GT:AD:DP:RGQ	0/0:2,0:2:6	./.:0,0:0	0/0:2,0:2:6
+20	10012734	.	G	.	0.63	LowQual	DP=4;ExcessHet=3.01;MQ=NaN	GT:DP:RGQ	0/0:2:6	0/0:0:0	0/0:2:6
```
Comment (ldgauthier) [inline src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf]: I looked at the input for this and there were PLs for all the samples (and a bunch more FORMAT annotations):
`20	10012751	.	T	C,<NON_REF>	.	.	DP=4;ExcessHet=3.01;RAW_MQ=14400.00	GT:AD:DP:GQ:MIN_DP:PL:SB	./.:0,2,0:2:6:.:49,6,0,49,6,49:0,0,1,1	./.:.:0:0:0:0,0,0,0,0,0	./.:0,2,0:2:6:.:49,6,0,49,6,49:0,0,1,1`
Specifically that first sample should be homVar (1/1) even though it's GQ is pretty crappy.  If they're all actually homRef then it would be very strange to call a variant there, even if it's very low quality.  Am I looking at the wrong input?

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/testSpanningDeletion.vcf [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+##INFO=<ID=RAW_MQ,Number=1,Type=Float,Description="Raw data for RMS Mapping Quality">
+##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias">
+##INFO=<ID=SOR,Number=1,Type=Float,Description="Symmetric Odds Ratio of 2x2 contingency table to detect strand bias">
+##contig=<ID=20,length=63025520,assembly=b37>
+##contig=<ID=21,length=48129895,assembly=b37>
+##reference=file:///Users/cnorman/projects/gatk/src/test/resources/large/human_g1k_v37.20.21.fasta
+##source=GenotypeGVCFs
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	GTEX-OXRP-0003	GTEX-QXCU-0004	GTEX-RVPV-0003
+20	10096905	.	TA	T	69.37	.	AC=2;AF=1.00;AN=2;DP=2;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=34.69;SOR=0.693	GT:AD:DP:GQ:PGT:PID:PL	1/1:0,2:2:6:1|1:10096899_G_T:90,6,0	./.:0,0:0:.:.:.:0,0,0	./.:0,0:0:.:.:.:0,0,0
+20	10096907	.	A	.	18.38	.	DP=2	GT:AD:DP	./.:2,0:2	./.:0,0:0	./.:0,0:0
```
Comment (ldgauthier) [inline src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/testSpanningDeletion.vcf]: I'm really happy that all the positions requested have data (905 and 906 in the deletion and 907 subsequently).  This does the right thing by dropping the *-only line in the combined GVCF.  Can you add output for 20: 10624924 -10624926, which has a spanning deletion with another alt allele?

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf (hunk 24 above)
Comment (ldgauthier) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf]: Missing ADs are also why the MQ=NaN.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java (hunk 5 above)
Comment (ldgauthier) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: I would rather call that in GATK4 as well.  If it's not variant then we don't need annotations for filtering downstream.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java (hunk 13 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: @ldgauthier GATK4 isn't adding `AN` because `ChromosomeCounts` isn't `Reducible` (so its [bypassed by the annotation engine here](https://github.com/broadinstitute/gatk/blob/67f0f0f2e59185b721398b17c24eba487a2ac76c/src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java#L217)). Perhaps thats a separate problem ?

File: src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java (hunk 14 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java]: I think I would change it.  It's more clear that way.

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/testSpanningDeletion.vcf (hunk 25 above)
Comment (cmnbroad) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/testSpanningDeletion.vcf]: Yeah - good idea.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java (hunk 13 above)
Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: But in your testSpanningDeletion output, there's an AC=2 and AN=2 on the deletion, which must have come from ChromosomeCounts.  Is it just not being called consistently?  We don't need it at non-variant sites.

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf (hunk 24 above)
Comment (ldgauthier) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf]: Totally looking at the wrong input.  Sorry.  I do wonder about the ADs, though it looks like that was also an issue in the GATK3 output.

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/testSpanningDeletion.vcf (hunk 25 above)
Comment (cmnbroad) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/testSpanningDeletion.vcf]: Done.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java (hunk 13 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: I updated the output files and removed the `AN` test exclusion.

**Comment by droazen** [comment]
@cmnbroad Looks like this needs a rebase

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java (hunk 9 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFConstants.java]: Wait, actually, we can't. `Allele.NON_REF_STRING` is the string for the symbolic allele ("<NON_REF>", whereas this is just the `NON_REF`. Allele doesn't seem to have just the string.

**Comment by gubrins** [comment]
I'm a bit lost, how can we apply this?

**Comment by lbergelson** [comment]
@gubrins if you have a question about how something works a 3 year PR probably isn't the best place to ask :)

We have a [forum](https://gatk.broadinstitute.org/hc/en-us/community/topics) for asking "how do I... ?" questions.
